ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/ASTConverter.java

```diff
@@ -577,6 +588,77 @@ private QueryBlockInfo convertSource(RelNode r) throws CalciteSemanticException
         ast = ASTBuilder.subQuery(left, sqAlias);
         s = new Schema((Union) r, sqAlias);
       }
+    } else if (r instanceof HiveTableFunctionScan &&
+        !canOptimizeOutLateralView((HiveTableFunctionScan) r)) {
```

Comment:
Is this optimization something that we could defer to a follow-up ticket so that we converge and merge this PR faster ? That would also make the diff and git history easier to follow.

If we want to push it here then let me share what I was thinking with a concrete SQL example grabbed from existing tests.
```sql
SELECT myCol from tmp_pyang_lv LATERAL VIEW explode(array(1,2,3)) myTab as myCol limit 3
```
The current plan is the following:
```
CBO PLAN:
HiveSortLimit(fetch=[3])
  HiveProject(mycol=[$5])
    HiveTableFunctionScan(invocation=[explode(ARRAY(1, 2, 3))], rowType=[RecordType(VARCHAR(2147483647) inputs, BIGINT BLOCK__OFFSET__INSIDE__FILE, VARCHAR(2147483647) INPUT__FILE__NAME, RecordType(BIGINT writeid, INTEGER bucketid, BIGINT rowid) ROW__ID, BOOLEAN ROW__IS__DELETED, INTEGER mytab.mycol)])
      HiveTableScan(table=[[default, tmp_pyang_lv]], table:alias=[tmp_pyang_lv])
```
I was thinking that we could have a CBO optimization that turns the plan above to something like the following:
```
CBO PLAN:
HiveSortLimit(fetch=[3])
  HiveTableFunctionScan(invocation=[explode(ARRAY(1, 2, 3))], rowType=[RecordType(INTEGER mytab.mycol)])
```

Reply:
I've made some changes for the plan which now includes a lateral function definition.